### DISCUSSION TITLE
The question of async session.add()

### DISCUSSION THREAD

**Discussion by miss85246** [sqlalchemy/sqlalchemy#8030] [Usage Questions] [answered]
Hi, I use sanic create a api server and I want get  model id after `session.add()` . 
This is my  useage：

1. create engine after server start :

```python
@server_app.listener("after_server_start")
async def server_init(_, __):
    server_app.ctx.database = await DBClient(**server_app.config.DB_CONFIG)
````

2. create session in request middleware:

```python
  async def create_database_session_middleware(request):
      request.app.ctx.database.session = request.app.ctx.database.maker()
```

3. add new User model to session:

```python
  async def create_data(self, data: Union[Any, List[Any]]) -> Union[Any, List[Any]]:
      try:
          for item in data if isinstance(data, list) else [data]:
              self.session.add(item)
          await self.session.commit()
          return data
      except Exception as e:
          await self.session.rollback()
          raise e
```

4. close session in  response middleware:

```python
async def close_database_session_middleware(request, response):
    await request.app.database.session.close()
```

It's looks very well. But when I test it , I got this exception:
```
greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place?
```

I'm already try create new session in `create_data` like this：

```python
    async def create_data(self, data: Union[Any, List[Any]]) -> Union[Any, List[Any]]:
        try:
            async with self.maker() as session:
                for item in data if isinstance(data, list) else [data]:
                    session.add(item)
                    await session.flush()
                await session.commit()
                return data
        except Exception as e:
            await self.session.rollback()
            raise e
```

but this way will get another error: 

```python
 Instance <Resource at 0x10780f130> is not bound to a Session; attribute refresh operation cannot proceed 
```

Is there any good idea? looking forward to your reply， Thanks

**Comment by zzzeek** [comment, marked as answer]
hi -

when using asyncio, as well as separately when using detached objects as seems to be the case here, you must set [expire_on_commit](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=expire_on_commit#sqlalchemy.orm.Session.params.expire_on_commit) to False. 

the asyncio [documentation](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#synopsis-orm) sets this parameter in all the ORM examples.  For general techniques to avoid implicit IO, see the steps at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession .

**Reply by miss85246** [reply]
thanks for your answer. I was resolved it use `session.refresh()`, this is code:
```python
async def create_data(self, data: Union[Any, List[Any]]) -> Union[Any, List[Any]]:
    for item in data if isinstance(data, list) else [data]:
        try:
            self.session.add(item)
            await self.session.commit()
            await self.session.refresh(item)
        except Exception as e:
            await self.session.rollback()
            raise e
    return data
```
